### PR TITLE
Add typed array conversion constructor

### PR DESCRIPTION
**Pull request by jordigulley**
Adds conversion syntax for typed arrays
# Example
```gdscript
# Creates a new float array
# float_arr is inferred as Array[float]
var float_arr = Array[float]([2.2, 30.1, 40.2]);
var converted_array = Array[int](float_arr); # [2, 30, 40]
```

The syntax is shown in the documentation as `Array[Variant]`:
image

# Conditions
1: The analyzer checks if the types provided are compatible:
image
If the base array is not fixed-typed, the user will receive a non-fatal error at runtime for any values that could not be converted.

2: This syntax only works for built-in types *and* if the type provided implements a typed-constructor `base, builtin_type, class_name, script`:
image

# Implementation
This is implemented by checking if a call is being made on a Subscript node, and if the subscript's identifier is a built-in type. The subscript's index is then used as the type for the constructor, and the type's information (`builtin_type`, `class_name`, `script`) is then implicitly provided to the method. This way, the constructor can only be used with classes that provide a typed-constructor implementation. ex: `Array(base, builtin_type, class_name, script)`

### CODE REVIEW

**Comment by jordigcs** [comment]
This implementation can be easily extended to future typed containers by just implementing a constructor with a signature of `base, builtin_type, class_name, script` (base, type information)

**Comment by adamscott** [comment]
CC @vnen @vonagam @anvilfolk

**Comment by jordigcs** [comment]
Linux test is failing because of an unassigned pointer, fixing and doing further testing now.

**Comment by jordigcs** [comment]
Fixed!

**Comment by adamscott** [comment]
@YuriSizov What do you think of the `Array[...]`?

I think we should maybe write it as `Array[Variant]`, or maybe `Array[<type>]`?

**Comment by adamscott** [comment]
> This syntax (Array[...](base)) is actually just an alias for Array(base, builtin_type, class_name, script).

We should document that this is in fact only syntactic sugar, for ease of use.

I don't think we should mask the error or lie to the user, otherwise. I think we should maybe think about doing this if it turns out that a lot of users are confused about the error, and a lot of complains crop up. Then, we could begin to imagine a solution.

But that's just my humble opinion about this.

**Comment by YuriSizov** [comment]
> What do you think of the `Array[...]`?
> 
> I think we should maybe write it as `Array[Variant]`, or maybe `Array[<type>]`?

I'm not sold on any of them. The current version is not very telling: ellipses are not used in this manner in any syntax, so it's not clear what this notation implies. `Array[Variant]` is purely misleading, suggesting it's a typed array of Variant. `Array[<type>]` and variations are closest to truth but also not something you can typically see in the Godot documentation. Maybe `Array[T]` would be at least somewhat familiar?

We could also add a new qualifier, "typed", and use it here. It'll display a short tooltip explaining what that means.

**Comment by YuriSizov** [comment]
> `Array[Variant]` is purely misleading, suggesting it's a typed array of Variant.

On a second thought, how would we notate that method if it was written in GDScript? What would we put after `->`? For methods that return different types we would say `Variant`, so maybe `Array[Variant]` is not as misleading as I initially assumed.

**Comment by jordigcs** [comment]
The "typed" qualify is already added as "typed_constructor" This could easily be changed to typed. I'll try that and see if that works better.
I agree that `Array[T]` would be the most familar, with a hint that elaborates on what the `T` stands for

**Comment by jordigcs** [comment]
Areay[Variant] wouldn't technically be correct for the return type, since it would be limited to the type provided in the constructor, in my opinion

**Comment by jordigcs** [comment]
> > This syntax (Array[...](base)) is actually just an alias for Array(base, builtin_type, class_name, script).
> 
> We should document that this is in fact only syntactic sugar, for ease of use.
> 
> I don't think we should mask the error or lie to the user, otherwise. I think we should maybe think about doing this if it turns out that a lot of users are confused about the error, and a lot of complains crop up. Then, we could begin to imagine a solution.
> 
> But that's just my humble opinion about this.

Sounds good, I'll make the necessary changes asap.

**Comment by adamscott** [comment]
> Areay[Variant] wouldn't technically be correct for the return type, since it would be limited to the type provided in the constructor, in my opinion

But as @YuriSizov explained, this does work:

```gdscript
extends Node

func _ready() -> void:
    print(test())

func test() -> Array[Variant]:
    var arr: Array[String] = ["Array", "of", "strings"]
    return arr
```

**Comment by jordigcs** [comment]
> > Areay[Variant] wouldn't technically be correct for the return type, since it would be limited to the type provided in the constructor, in my opinion
> 
> But as @YuriSizov explained, this does work:
> 
> ```gdscript
> extends Node
> 
> func _ready() -> void:
>     print(test())
> 
> func test() -> Array[Variant]:
>     var arr: Array[String] = ["Array", "of", "strings"]
>     return arr
> ```
> 
> I'm in favor of typing as `Array[Variant]`. No matter what, it's the first time that it's considered a type in the docs, I think.
> 
> `T` makes me think of generics, which GDScript doesn't support.

Makes sense!

**Comment by jordigcs** [comment]
@YuriSizov Would a qualifier titled "shorthand" or "alias" also be a good idea? I think it would help make it more explicit that this is a shorthand for a different constructor, without them having to move down the description of the constructor.

**Comment by YuriSizov** [comment]
> Would a qualifier titled "shorthand" or "alias" also be a good idea? I think it would help make it more explicit that this is a shorthand for a different constructor, without them having to move down the description of the constructor.

We don't normally have aliases for methods in the API, it is strictly frowned upon as part of the design goals. So I wouldn't add "alias". As for shorthand, I don't think it would be much of a help, as you still need to read the description to understand what is it a shorthand for and how is it mapped to the other method.

**Comment by jordigcs** [comment]
Alright, thanks!

**Comment by vonagam** [comment]
> `T` makes me think of generics, which GDScript doesn't support.

Isn't usage of `T` to denote single type parameter is a standardized thing for languages? 
It will be not a Variant inside brackets but a type specifier.

**Comment by jordigcs** [comment]
Updated this PR with the following changes:
* "typed" qualifier that is applied to builtin constructors that accept a type.
* Typed constructors will have `[Variant]` applied to the return type and the method name in documentation, with a hint describing it's purpose.
* Moved type constructor logic completely into `parse_call` instead of separating it across `parse_subscript`
 * Allow narrowing conversion in array conversions, with a warning.
 
 I still need to add documentation for the typed constructor alias, but I'll do that after work.

File: core/variant/array.cpp [hunk 1]
```diff
@@ -193,8 +194,12 @@ uint32_t Array::recursive_hash(int recursion_count) const {
 
 bool Array::_assign(const Array &p_array) {
 	bool can_convert = p_array._p->typed.type == Variant::NIL;
-	can_convert |= _p->typed.type == Variant::STRING && p_array._p->typed.type == Variant::STRING_NAME;
-	can_convert |= _p->typed.type == Variant::STRING_NAME && p_array._p->typed.type == Variant::STRING;
+	can_convert |= Variant::can_convert_strict(p_array._p->typed.type, _p->typed.type);
+
+	bool narrowing_conversion = _p->typed.type == Variant::INT && p_array._p->typed.type == Variant::FLOAT;
+	if (narrowing_conversion) {
+		WARN_PRINT("Narrowing conversion in array assignment (float is converted to int and loses precision).");
```
Comment (jordigcs) [inline core/variant/array.cpp:201]: This warning could also be added to the analyzer, since it's checking if the types are compatible anyways when you use `Array[T](base)`

**Comment by adamscott** [comment]
> * Typed constructors will have `[Variant]` applied to the return type and the method name in documentation, with a hint describing it's purpose. Example:

Well, if there's the new `typed` tag, maybe `Array[T]` is not so bad, after all.

**Comment by vonagam** [comment]
Important to point out that current system by which type of array set is quite limiting one. To set the type 3 arguments need to be passed around from analyzer to compiler to vm to array. It also does not support nested typing - `Array[Array[int]]` - even though there is nothing fundamentally stopping it from being implemented. And next builtin types that can theoretically have parameters - `Dictionary` and `Callable` - both ask for two parameters - key/value and args/return. (At that point in the future I think it would make sense to maybe introduce new class in codebase that will uniquely represent a type and have properties of `builtin type, native type, script type, parameters`.) 

I'm saying that just so that there should not be too much focus on implementing generic solution for other types with assumption that they have only one parameter and the same 3 arguments will be used for them.

**Comment by jordigcs** [comment]
I also believe making types a first class object would be good, but I'm not sure if it'd have many use cases. That sounds interesting, if others would like to give their input on that :)

**Comment by adamscott** [comment]
@jordigcs Is it ready to be merged?

**Comment by jordigcs** [comment]
This is ready, I'll add tests. Also, I'm not sure why the current tests are failing. This PR doesn't change anything about read-only arrays... Is this a bug with the test?

**Comment by anvilfolk** [comment]
It's a known problem :) I believe @vonagam will be fixing that soon :) It's an issue where runtime error messages are platform specific.

File: core/variant/array.cpp (hunk 1 above)
Comment (vnen) [reply core/variant/array.cpp:201]: Actually, I don't think it's necessary to warn here. We don't really have this sort of warnings for runtime. If anything, it should be debug only.

File: modules/gdscript/tests/scripts/analyzer/errors/typed_constructor_not_implemented.out [hunk 2]
```diff
@@ -0,0 +1,2 @@
+GDTEST_ANALYZER_ERROR
+No constructor of "float" matches the signature "float(Array, int, StringName, Variant)"
```
Comment (vnen) [inline modules/gdscript/tests/scripts/analyzer/errors/typed_constructor_not_implemented.out:2]: That's a confusing message, since this constructor format was not actually typed by the user. This should say only arrays can use this type of constructor since it's the only type that actually have type argument. If/when this we add more possibilities in the future we can change the message.

File: modules/gdscript/gdscript_parser.cpp [hunk 3]
```diff
@@ -2918,8 +2917,29 @@ GDScriptParser::ExpressionNode *GDScriptParser::parse_call(ExpressionNode *p_pre
...
+							type->type_chain.push_back(index);
+							reset_extents(type, index);
+							complete_extents(type);
+							call->typed = type;
+						} else {
+							push_error(R"*(Typed constructors can only be called with a valid Type.)*", index);
+						}
+					} else {
+						push_error(R"*(Cannot call on an expression. Use ".call()" if it's a Callable.)*");
+					}
```
Comment (vnen) [inline modules/gdscript/gdscript_parser.cpp:2938]: I feel like the parser is doing too much here. Checking it the type is valid type is the job of the analyzer. A type can be more than identifier, it might be an attribute access (`SomeClass.InnerClass`) or technically even a `preload`ed script.

The only thing the parser should check is if the base is `Array` since it's the only type that has this kind of constructor (can be changed later in the future). At most, check if it's a builtin type and give the error about typed constructors if it's builtin but not an array and if it's not builtin give the error about calling on an expression.

**Comment by vnen** [comment]
Looks mostly okay to me. I registered my concerns in the comments. Still will need to see with the rest of the core team if the changes are good.

**Comment by jordigcs** [comment]
Resolved, thanks!

**Comment by jordigcs** [comment]
Need to update tests.

**Comment by vnen** [comment]
This seems okay to me. It needs to fix the CI issues.

File: doc/classes/Array.xml [hunk 4]
```diff
@@ -62,6 +62,13 @@
 				Creates a typed array from the [param base] array.
 			</description>
 		</constructor>
+		<constructor name="Array" qualifiers="typed">
+			<return type="Array" />
+			<param index="0" name="base" type="Array" />
+			<description>
+				Alias for `Array(base, type, class_name, script)`. Implicitly fills type-related parameters from the type provided. Creates a typed array from the [param base] array.
```
Comment (akien-mga) [inline doc/classes/Array.xml:69]:
```suggestion
				Alias for [code]Array(base, type, class_name, script)[/code]. Implicitly fills type-related parameters from the type provided. Creates a typed array from the [param base] array.
```

But see also the `--doctool` output, it doesn't seem to support the two different `Array(Array)` constructors, it might need more code to include the `typed` qualifiers for disambiguation.

**Comment by ttencate** [comment]
> > Areay[Variant] wouldn't technically be correct for the return type, since it would be limited to the type provided in the constructor, in my opinion
> 
> But as @YuriSizov explained, this does work:
> 
> ```gdscript
> extends Node
> 
> func _ready() -> void:
>     print(test())
> 
> func test() -> Array[Variant]:
>     var arr: Array[String] = ["Array", "of", "strings"]
>     return arr
> ```

But this does _not_ work:

```gdscript
extends Node

func _ready() -> void:
    var arr: Array[Variant] = test()
    arr.push_back(42)  # Huh? Why can't I push an integer to a *variant* array?

func test() -> Array[Variant]:
    # ...
```

Arrays are not covariant: `Array[String]` is not a subtype of `Array[Variant]`.

I have no solution but just thought I'd mention this.

**Comment by YuriSizov** [comment]
@ttencate Well, that should work, because any value that you can create in GDScript is actually a variant to the engine. So the original point stands.

**Comment by vonagam** [comment]
> Arrays are not covariant 

That is correct.

> `Array[String]` is not a subtype of `Array[Variant]`.

`Array[Variant]` is an exception though - currently there is no typed array of `Variant`, even during the parsing `Variant` part is dropped, so it is just `Array`. And `Array` type has to accept other typed arrays, otherwise it will be impossible to use typed arrays with many built-in methods. For example `Array#append_array` expects `Array`, if `Array` would not allow a typed array as a value then appending typed arrays with this method will not work.

**Comment by YuriSizov** [comment]
I wouldn't even call it an exception. `Variant` is a type in the engine, but in GDScript it's only used to mean `Any`. While other languages can have their own data and container types and whatnot, GDScript is a part of Godot, and it only deals in variants. So we just use the term from the engine, where perhaps it would've been clearer to introduce a more unambiguous keyword to mean "any type".

**Comment by ttencate** [comment]
> @ttencate Well, that should work, because any value that you can create in GDScript is actually a variant to the engine. So the original point stands.

It doesn't actually work, because there's a runtime type check on typed arrays. Allowing `Array[String]` to be interpreted as `Array[Variant]` does not remove the type check. The write is ignored and you get an error logged at runtime:

```gdscript
extends Node

func _ready() -> void:
	var arr: Array[Variant] = test()
	arr.push_back(42)  # Huh? Why can't I push an integer to a *variant* array?

func test() -> Array[Variant]:
	var arr: Array[String] = ["Array", "of", "strings"]
	return arr
```

> E 0:00:00:0602   main.gd:5 @ _ready(): Attempted to push_back a variable of type 'int' into a TypedArray of type 'String'.
> E 0:00:00:0602   main.gd:5 @ _ready(): Condition "!_p->typed.validate(value, "push_back")" is true.

File: doc/classes/Array.xml (hunk 4 above)
Comment (jordigcs) [reply doc/classes/Array.xml]: Fixed, thanks!

**Comment by adamscott** [comment]
@vnen Are we still waiting for @reduz input?

**Comment by Mickeon** [comment]
So... what's the deal with this PR? Reception seems fairly positive

**Comment by nlupugla** [comment]
I'm quite late to the party here, but this PR was sent my way as I've been thinking a bit about how to deal with some of the issues with typed arrays in GDScript currently: https://github.com/godotengine/godot-proposals/discussions/7364.

The idea of TypedArray constructors seems like a great feature to me. I skimming over the discussion here, it seems like most of hit has been about the syntax and documentation, but I'm more curious about the semantics of the constructor. In the OP, @jordigcs mentions that constructing an array of an incompatible type results in a non-fatal error. Can you elaborate on that? For example, suppose I have

```gdscript
var ints: Array = [1, 3]
var vectors := Array[Vector2](ints) # non-fatal runtime error
print(vectors) # what happens here?
```
I could imagine a few possible outcomes of the print statement.
1. Another runtime error, nothing gets printed (except an error message of course)
2. null
3. An empty array
4. An array of "default" Vector2s, which I suppose is `Vector2.ZERO`.
5. An array of nulls

For comparison, I get a fatal error if I try this with single elements:
```gdscript
var my_int: Variant = 1
var vec: Vector2 = Vector2(int) # fatal error
print(vec) # never executes because of error
```

Overall, it would be nice to have a clear breakdown of all the edge cases and failure modes of these typed array constructors.

**Comment by RadiantUwU** [comment]
I'll test this later today and make sure it still works as expected.

**Comment by Mickeon** [comment]
Regardless of it, it would still need to be either rebased on be salvaged by someone else. This is _extremely_ promising, though.

**Comment by pewcworrell** [comment]
Just chiming in since it's been over 5 months to say this enhancement is still very much desired.

**Comment by produno** [comment]
Maybe its time this was assigned to someone on the team? Its been quite some time since typed arrays were implemented, but not having a proper constructor and having to create a new array and assign the old array to the new one, makes them still feel half finished.

Not only that but other 'features' keep being worked on whilst previously implemented features are still not fully fleshed out, like typed arrays.

**Comment by AThousandShips** [comment]
Would be best if someone takes it over and salvages it

**Comment by JoshuaWeissTBS** [comment]
Chiming in also, this is super desired. Any timeline on this?

**Comment by jordigulley** [comment]
> I'm quite late to the party here, but this PR was sent my way as I've been thinking a bit about how to deal with some of the issues with typed arrays in GDScript currently: [godotengine/godot-proposals#7364](https://github.com/godotengine/godot-proposals/discussions/7364).
> 
> The idea of TypedArray constructors seems like a great feature to me. I skimming over the discussion here, it seems like most of hit has been about the syntax and documentation, but I'm more curious about the semantics of the constructor. In the OP, @jordigcs mentions that constructing an array of an incompatible type results in a non-fatal error. Can you elaborate on that? For example, suppose I have
> 
> ```gdscript
> var ints: Array = [1, 3]
> var vectors := Array[Vector2](ints) # non-fatal runtime error
> print(vectors) # what happens here?
> ```
> 
> I could imagine a few possible outcomes of the print statement.
> 
> 1. Another runtime error, nothing gets printed (except an error message of course)
> 2. null
> 3. An empty array
> 4. An array of "default" Vector2s, which I suppose is `Vector2.ZERO`.
> 5. An array of nulls
> 
> For comparison, I get a fatal error if I try this with single elements:
> 
> ```gdscript
> var my_int: Variant = 1
> var vec: Vector2 = Vector2(int) # fatal error
> print(vec) # never executes because of error
> ```
> 
> Overall, it would be nice to have a clear breakdown of all the edge cases and failure modes of these typed array constructors.

This implementation is just an alias for [this constructor](https://docs.godotengine.org/en/stable/classes/class_array.html#class-array-constructor-array) Godot already has, in this constructor, the constructor just silently fails and returns an empty array. I think improving the error handling for this existing constructor might be outside of the scope for this PR.

File: editor/editor_help.cpp [hunk 5]
```diff
@@ -156,6 +156,8 @@ static void _add_qualifiers_to_rt(const String &p_qualifiers, RichTextLabel *p_r
 			hint = TTR("This method does not need an instance to be called.\nIt can be called directly using the class name.");
 		} else if (qualifier == "abstract") {
 			hint = TTR("This method must be implemented to complete the abstract class.");
+		} else if (qualifier == "typed") {
+			hint = TTR("This method can recieve a type.");
```
Comment (Mickeon) [inline editor/editor_help.cpp]:
```suggestion
			hint = TTR("This method can receive a type.");
```

File: modules/gdscript/tests/scripts/analyzer/warnings/typed_array_narrowing_conversion.gd [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# Typed conversion constructors are currently only implemented for Array
+# Update tests if other built-in types gain this functionality eg: Typed dictionarys are implemented
```
Comment (Mickeon) [inline modules/gdscript/tests/scripts/analyzer/warnings/typed_array_narrowing_conversion.gd]: You say this but the file name of the test explicitly says `typed_array`, so this is for granted.
```suggestion
# Typed conversion constructors are currently only implemented for Array.
```
I would suggest also removing this line from the other tests.

**Review by Mickeon — COMMENTED** [review]
Beware that unit tests failed because of typos.

File: modules/gdscript/tests/scripts/analyzer/warnings/typed_array_narrowing_conversion.gd (hunk 6 above)
Comment (jordigulley) [reply modules/gdscript/tests/scripts/analyzer/warnings/typed_array_narrowing_conversion.gd]: Got it!

**Comment by nlupugla** [comment]
> This implementation is just an alias for [this constructor](https://docs.godotengine.org/en/stable/classes/class_array.html#class-array-constructor-array) Godot already has, in this constructor, the constructor just silently fails and returns an empty array. I think improving the error handling for this existing constructor might be outside of the scope for this PR.

Thanks for coming back to this! I think it will be a very welcome addition.

I disagree that improving the error handling is beyond the scope of this PR. Godot's internal array constructor is acceptable in it's current state because it's not user facing. When exposing functionality to users, I think it is very much in-scope to bring the ergonomics and error-handling up to par. 

In fact, I would err on the side of being more restrictive (e.g. fatal error) initially because it's generally for the sake of compatibility to add options than remove them.

**Comment by jordigulley** [comment]
> > This implementation is just an alias for [this constructor](https://docs.godotengine.org/en/stable/classes/class_array.html#class-array-constructor-array) Godot already has, in this constructor, the constructor just silently fails and returns an empty array. I think improving the error handling for this existing constructor might be outside of the scope for this PR.
> 
> Thanks for coming back to this! I think it will be a very welcome addition.
> 
> I disagree that improving the error handling is beyond the scope of this PR. Godot's internal array constructor is acceptable in it's current state because it's not user facing. When exposing functionality to users, I think it is very much in-scope to bring the ergonomics and error-handling up to par.
> 
> In fact, I would err on the side of being more restrictive (e.g. fatal error) initially because it's generally for the sake of compatibility to add options than remove them.


It turns out Array actually already has error handling for invalid items. It still returns an empty array, but it prints an error to the console listing which items were invalid. I think the docs should clarify this.

```
    var v = Array[Node2D]([3, 4, 5]);
    print(v);
```
image
